versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/ReferenceLogicImpl.java

```diff
@@ -0,0 +1,644 @@
+      @Nonnull @jakarta.annotation.Nonnull ObjId expectedPointer)
+      throws RefNotFoundException, RefConditionFailedException, RetryTimeoutException {
+    checkArgument(!isInternalReferenceName(name));
+
+    Reference reference = persist.fetchReference(name);
+    if (reference == null) {
+      StoreKey nameKey = key(name);
+      Supplier<StoreIndex<CommitOp>> indexSupplier = createRefsIndexSupplier();
+      StoreIndexElement<CommitOp> index = indexSupplier.get().get(nameKey);
+      if (index == null) {
+        // not there --> okay
+        throw new RefNotFoundException(reference(name, expectedPointer, false));
+      }
+      reference = maybeRecover(name, null, indexSupplier);
+
+      if (reference == null) {
+        // not there, even after recovery
+        throw new RefNotFoundException(reference(name, expectedPointer, false));
+      }
+    }
+
+    boolean actAsAlreadyDeleted = reference.deleted();
+
+    if (!reference.pointer().equals(expectedPointer)) {
+      // A previous deleteReference failed, act as if the first one succeeded, therefore this
+      // one must throw a ReferenceNotFoundException instead of a ReferenceConditionFailedException
+      if (!actAsAlreadyDeleted) {
+        Supplier<StoreIndex<CommitOp>> indexSupplier = createRefsIndexSupplier();
+        Reference recovered = maybeRecover(name, reference, indexSupplier);
+        throw new RefConditionFailedException(recovered != null ? recovered : reference);
+      }
+    }
+
+    if (!actAsAlreadyDeleted) {
+      persist.markReferenceAsDeleted(reference);
+    }
+
+    commitDeleteReference(reference);
+
+    persist.purgeReference(reference);
+
+    if (actAsAlreadyDeleted) {
+      // A previous deleteReference failed, act as if the first one succeeded, therefore this
+      // one would have not found the reference.
+      throw new RefNotFoundException(reference(name, expectedPointer, false));
+    }
+  }
+
+  static final class CommitReferenceResult {
+    final Reference reference;
+    final Kind kind;
+
+    CommitReferenceResult(Reference reference, Kind kind) {
+      this.reference = reference;
+      this.kind = kind;
+    }
+
+    enum Kind {
+      ADDED_TO_INDEX,
+      REF_ROW_EXISTS,
+      REF_ROW_MISSING
+    }
+  }
+
+  @VisibleForTesting // needed to simulate recovery scenarios
+  CommitReferenceResult commitCreateReference(String name, ObjId pointer)
+      throws RetryTimeoutException {
+    try {
+      return commitRetry(
+          persist,
+          (p, retryState) -> {
+            Reference refRefs = requireNonNull(p.fetchReference(REF_REFS.name()));
+            long created = p.config().currentTimeMicros();
+            RefObj ref = ref(name, pointer, created);
+            ObjId refObjId;
+            try {
+              refObjId = p.storeObj(ref);
+              if (refObjId == null) {
+                // Oops, such a RefObj already exists
+                throw new RetryException();
```

Comment:
It looks like this try will run with the same input data and hit the same object ID collision (we do not expect objects to be removed from the store, right?)

Reply:
handles races (concurrent, same ref-creation)

Reply:
maybve - it likely never ever happens

Reply:
Since `RefObj` is immutable, do we have to re-try in this case? Why not simply carry on?

Reply:
feels also doable - probably better.